### PR TITLE
Initial commit for review before tagging for release

### PR DESCRIPTION
**Pull request by john-n-smith**
https://github.com/graze/telnet-client/blob/initialCommit/README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,151 @@
...
+$client = Graze\TelnetClient\TelnetClient::build($dsn);
+$command = 'Open the pod bay doors, HAL';
+$resp = $client->execute($command);
+```
+
+The `execute` method returns a `TelnetResponse` object:
+
+```php
+/* @return bool */
+$resp->getIsError(); // whether an error prompt was encountered
```
Comment (wpillar) [inline README.md]: `->isError()` ?

File: src/InterpretAsCommand.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * @var string
+     */
+    protected $DO;
+
+    /**
+     * @var string
+     */
+    protected $DONT;
```
Comment (wpillar) [inline src/InterpretAsCommand.php:40]: Why the capitalisation? Seems a bit at odds to how we usually do variables.

File: src/TelnetClient.php [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+    /**
+     * @var string
+     */
+    protected $DC1;
+
+    /**
+     * @var string
+     */
+    protected $IAC;
```
Comment (wpillar) [inline src/TelnetClient.php:73]: Same about the capitalisation?

File: src/TelnetClientBuilder.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        return new InterpretAsCommand();
+    }
+
+    /**
+     * @return TelnetClient
+     */
+    public function getTelnetClient()
+    {
+        return new TelnetClient();
+    }
```
Comment (wpillar) [inline src/TelnetClientBuilder.php:112]: A `TelnetClientBuilder` has public methods for constructing other objects? Seems odd.

File: src/TelnetResponse.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    protected $promptMatches;
+
+    /**
+     * @param bool $isError
+     * @param string $responseText
+     * @param array $promptMatches
+     */
+    public function __construct($isError, $responseText, array $promptMatches)
+    {
+        $this->isError = $isError;
```
Comment (wpillar) [inline src/TelnetResponse.php]: Might be worth a `(bool)` or might be too defensive.

File: src/InterpretAsCommand.php (hunk 2 above)
Comment (john-n-smith) [inline src/InterpretAsCommand.php:40]: Ideally they would be constants, but need to be initialised by a method call so trying to highlight that. Happy to go lc though,

File: src/TelnetClientBuilder.php (hunk 4 above)
Comment (john-n-smith) [inline src/TelnetClientBuilder.php:112]: Allows me to intercept and mock in order to test the builder. Open to suggestions.

Comment (wpillar) [inline src/TelnetClientBuilder.php:112]:
``` php
class TelnetClientBuilder
{
    private $socketFactory;

    public function  __construct(SocketFactory $socketFactory)
    {
        $this->socketFactory = $socketFactory;
    }

    public static function factory()
    {
        return new static(new SocketFactory());
    }

    public function build()
    {
        // do something with $this->socketFactory
    }
}
```

They're dependencies, so I'd inject them into the builder. Then add a convenient `::factory()` method to help users out.

Comment (john-n-smith) [inline src/TelnetClientBuilder.php:112]: There's slightly more untestable logic in the `factory` method than the original approach but I prefer  having fewer functions.

File: src/TelnetResponse.php (hunk 5 above)
Comment (john-n-smith) [inline src/TelnetResponse.php]: No harm in adding it.

File: src/TelnetClientBuilder.php (hunk 4 above)
Comment (wpillar) [inline src/TelnetClientBuilder.php:112]: What becomes untestable? Some `new` statements?

**Comment by john-n-smith** [comment]
@wpillar, are you happy to mark this as approved?

**Comment by wpillar** [comment]
@john-n-smith what about the scrutinizer issues?

**Comment by john-n-smith** [comment]
@wpillar all that work then forget to use it...
All the actual bugs are fixed. There was a couple of false alarms and some style suggestions that are not in keeping with our standards.

**Comment by wpillar** [comment]
@john-n-smith cool, I'm happy. Merge away.

**Pull request merged**